### DISCUSSION TITLE
Matplotlib plots

### DISCUSSION THREAD

**Discussion by csteele2** [UXARRAY/uxarray#839] [Q&A] [answered]
I have a couple of questions that maybe others have come across. While the gallery of the uxarray documentation shows figures with different projections, when I use the same code, I get different looking figures, presumably because the image returned seems to have a regular axes, not a geoaxes.

My other issue is for real plots. Interactive plots in a notebook are nice, but what if I wanted to make scripted png plots? As far as I have tried, I can not extract a matplotlib figure or axes, as only an image gets returned. Is there a way to plot the data directly in matplotlib without using the built-in .plot? I would like to have more, easier, control over the axes and figure than seems possible using the built-in functions and not have the holoviews,etc. dependecies on a headless server. Additionally, a contour function would very much be needed, but I don't see much on that. There are some tangential threads on this already, but in order to really use MPAS seriously, I need to be able to make similar plots as a WRF workflow. I don't consider interpolating the data back to a regular grid to be more than a stopgap, as that has high potential to introduce error (especially with any sort of station verification). Happy to add any portion of this to the prioritizations thread if appropriate.

**Comment by philipc2** [comment, marked as answer]
Hi @csteele2 

Our direct `matplotlib` support is very minimal as of right now. We have some support for creating `PolyCollection` and `LineCollection` objects that can be used in within `matplotlib` workflows, however the performance is much less than ideal and we haven't spend much time recently looking into it. However, this is something I can take a look at this upcoming iteration, since it does seem like someone the community would benefit from.

We can track the progress with #840 and #430 

[This page might be a good read from HoloViews.](https://holoviews.org/user_guide/Plotting_with_Matplotlib.html#working-with-matplotlib-directly). All of our `.plot()` methods suppot a `backend='matplotlib'` parameter, which uses matplotlib as the renderer through HoloViews.

**Reply by csteele2** [reply]
Okay, that looks like a decent band aid. So, the only plot method that will return a geoaxes is points? when trying like polygons, I notice it doesn't make a geoaxis, and even the rendered image is just a simple lat, lon image.

**Comment by csteele2** [comment]
Okay, I was able to pull the matplotlib elements out of holoviews and use them, albeit with some difficulties remaining when a geoaxes isn't returned.

**Comment by philipc2** [comment]
@csteele2 

We've added better support for directly using Matplotlib in our most recent release. We've added a user guide section describing how to generate plots, including projection and geographic feature support

https://uxarray.readthedocs.io/en/latest/user-guide/mpl.html

Also pinging @rljacob @brianpm @ahijevyc if you have any input on it!

**Reply by ahijevyc** [reply]
I love the new matplotlib support--especially cartopy. I ran the examples in the user guide with my own data.
